From 1dfca3f45f9ad5dfd7e22e899bcacac54affdec6 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Sun, 12 Sep 2021 13:11:06 +0300
Subject: [PATCH] [SW-55214] ppboot: gaudi2: enable axi drain

AXI drain is being enabled early in the boot sequence, so if the
code is accessing a block that is still held in reset state (like
pci blocks during prstn) the transaction will be handled and not
cause a device hanging.

Change-Id: I343002056f74bcc93f5a15ff0c12e022a57bb0dc
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 81a1c72..67cec60 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -114,6 +114,8 @@
 	/* Functionality used to be done in BTL */
 	init_prstn();
 
+	gaudi2_axi_drain_ena();
+
 	init_secondary();
 
 	/* This W/A is to be done before pcie link is established */
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index f780dfa..f9bf398 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -113,6 +113,8 @@
 {
 	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
 
+	gaudi2_axi_drain_ena();
+
 	/* This W/A is to be done before pcie link is established */
 	set_pci_msix_pba_offset_val();
 }
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 219e9b7..b0e547c 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -701,3 +701,11 @@
 	val &= ~mask;
 	regmap_writel(val, mmPSOC_GPIO3_DATAOUT);
 }
+
+#define AXI_DRAIN_TIMEOUT	0x1000
+
+void gaudi2_axi_drain_ena(void)
+{
+	regmap_writel(AXI_DRAIN_TIMEOUT, mmPSOC_GLOBAL_CONF_AXI_DRAIN_TIMEOUT);
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
+}
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index f428cf8..76a6ffc 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -224,3 +224,5 @@
 	unsigned int blocks;
 	unsigned int instances;
 };
+
+void gaudi2_axi_drain_ena(void);
